src/2025h2/autoreborrow-traits.md

```diff
@@ -0,0 +1,139 @@
+user can define this type as having exclusive semantics by not making the type `Copy`, but they cannot opt
+into automatic reborrowing. The best they can hope is to implement a custom `.reborrow_mut()` method similar
+to the `Option::as_deref_mut` from above. Here again they run into the issue that the lifetime of a
+`Mut<'_, T>` always gets constrained to the closes `.reborrow_mut()` call, making it impossible to return
+values derived from a `Mut<'_, T>` from the function that called `.reborrow_mut()`.
+
+An improvement is needed.
+
+### The next 6 months
+
+- Bring up an RFC for autoreborrowing: a
+  [draft](https://github.com/aapoalas/rfcs/blob/autoreborrow-traits/text/0000-autoreborrow-traits.md) exists.
+- Choose the internal logic of recursive reborrowing: based on core marker types, or on interaction with
+  `Copy`?
+- Implement nightly support for non-recursive reborrowing.
+- Gather feedback from users, especially `reborrow` crate users.
+- Implement nightly support for recursive reborrowing.
+
+### The "shiny future" we are working towards
+
+`Pin` ergonomics group should be able to get rid of special-casing of `Pin` reborrowing in rustc.
+
+Rust for Linux project should be enabled to experiment with custom reborrowable reference types in earnest.
+
+Users of `reborrow` crate and similar should be enabled to move to core solution.
+
+## Design axioms
+
+- Accept the definition of reborrowing as "a memory copy with added lifetime analysis".
+  - This disallows running user code on reborrow.
+  - "Reborrow-as-shared" likely needs to run user code; this'd preferably be ignored where possible.
+- Must achieve true reborrowing, not a fascimile.
+  - Current userland reborrowing uses `T::reborrow_mut` functions that achieve the most important part of
+    reborrowing, temporarily disabled `T` upon reborrow.
+  - Userland cannot achieve true reborrowing: true reborrowing does not constrain the lifetime of `T`,
+    whereas userland fascimile does.
+  - Difference is in whether values derived from a reborrow can be returned past the point of the reborrow.
+- Performance of the solution must be trivial: reborrow is checked for at every coercion site. This cannot be
+  slow.
+- Make sure autoreborrowing doesn't become a vehicle for implicit type coercion. Allowing autoreborrowing
+  from `T` to multiple values could be abused to define a `CustomInt(int128)` that coerces to all integer
+  types.
+  - Autoreborrow traits should use an associated type instead of a type parameter.
+  - Autoreborrowing at coercion sites should not dovetail into eg. an `Into::into` call.
+
+## Ownership and team asks
+
+| Task                         | Owner(s) or team(s) | Notes |
+|------------------------------|---------------------|-------|
+| Discussion and moral support | ![Team][] [lang]    | Normal RFC process |
+| Help around impl             | ![Team][] [compiler]| Trait-impl querying in rustc to replace `Pin<&mut T>` special case |
```

Comment:
I anticipate this being slow, which is one of the reasons for my suggestion below to use an attribute on fields.

Reply:
Is it slow for rustc to query a type `T` for if `T: Trait`? Currently there's a check for if a `T` is an ADT and it matches the `Pin` type, and contains `&mut T` inside it. If this is true, then a custom autoreborrow adjustment is injected at a coercion place.

Autoreborrow traits would change that check to be "if `T` is an ADT and it implements `T: Reborrow`". Checking fields of `T` or anything like that wouldn't come into it at all.

Reply:
(also: to pass CI

```suggestion
| Standard reviews             | ![Team][] [compiler]| Trait-impl querying in rustc to replace `Pin<&mut T>` special case |
```
)